File: src/python/xerxes/context/compaction_provisioner.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Agent-backed context compaction and pre-append provisioning."""

from __future__ import annotations

import json
import logging
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

from .token_counter import SmartTokenCounter

logger = logging.getLogger(__name__)

COMPACTION_SUMMARY_PREFIX = "[Previous conversation summary"
_DEFAULT_THRESHOLD_RATIO = 0.75
_DEFAULT_TARGET_RATIO = 0.5
_DEFAULT_TAIL_RATIO = 0.35
_LOCAL_PROVIDER_NAMES = {"custom", "lmstudio", "ollama"}

CompactionSummaryAgent = Callable[[list[dict[str, Any]], str | None], str]


@dataclass(frozen=True)
class CompactionProvision:
    """Result from one compaction provisioning pass."""

    compacted: bool
    messages: list[dict[str, Any]]
    tokens_before: int
    tokens_after: int
    summarized_count: int = 0
    kept_count: int = 0
    reason: str = ""
    error: str = ""


def message_content_to_text(content: Any) -> str:
    """Return a readable text rendering for provider message content."""
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts = []
        for item in content:
            if isinstance(item, dict):
                parts.append(str(item.get("text") or json.dumps(item, ensure_ascii=False, default=str)))
            else:
                parts.append(str(item))
        return "\n".join(parts)
    if isinstance(content, dict):
        return json.dumps(content, ensure_ascii=False, default=str)
    return str(content or "")


def render_messages_for_summary(messages: list[dict[str, Any]]) -> str:
    """Render messages as complete compaction-agent input."""
    rendered = []
    for index, message in enumerate(messages, start=1):
        role = str(message.get("role", "unknown")).upper()
        lines = [f"Message {index} [{role}]"]
        content = message_content_to_text(message.get("content", ""))
        if content:
            lines.append(content)
        if message.get("tool_calls"):
            lines.append("tool_calls=" + json.dumps(message.get("tool_calls"), ensure_ascii=False, default=str))
        if message.get("tool_call_id"):
            lines.append(f"tool_call_id={message.get('tool_call_id')}")
        rendered.append("\n".join(lines))
    return "\n\n".join(rendered)


def _tool_call_names(message: dict[str, Any]) -> dict[str, str]:
    """Return ``tool_call_id -> tool name`` for an assistant message."""
    names: dict[str, str] = {}
    for tool_call in message.get("tool_calls") or []:
        tool_call_id = tool_call.get("id")
        if tool_call_id:
            names[str(tool_call_id)] = str(tool_call.get("name") or "")
    return names


def _complete_pending_tool_results(out: list[dict[str, Any]], pending: dict[str, str]) -> None:
    """Append synthetic tool results for assistant calls that lost their result."""
    for tool_call_id, name in pending.items():
        out.append(
            {
                "role": "tool",
                "tool_call_id": tool_call_id,
                "name": name,
                "content": "[Tool result unavailable after context compaction]",
                "is_error": True,
            }
        )
    pending.clear()


def repair_tool_message_sequence(messages: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Return messages with provider-valid assistant/tool-call ordering.

    OpenAI-compatible providers reject a ``tool`` message unless it answers
    a preceding assistant ``tool_calls`` entry. Compaction changes the leading
    edge of the request window, so orphan tool results are dropped and missing
    recent tool results are represented by synthetic error messages.
    """
    out: list[dict[str, Any]] = []
    pending: dict[str, str] = {}

    for message in messages:
        role = message.get("role")
        if role == "assistant":
            if pending:
                _complete_pending_tool_results(out, pending)
            out.append(message)
            pending = _tool_call_names(message)
            continue

        if role == "tool":
            tool_call_id = str(message.get("tool_call_id") or "")
            if tool_call_id and tool_call_id in pending:
                out.append(message)
                pending.pop(tool_call_id, None)
            continue

        if pending:
            _complete_pending_tool_results(out, pending)
        out.append(message)

    if pending:
        _complete_pending_tool_results(out, pending)
    return out


class ProviderCompactionAgent:
    """LLM-backed summary agent used when no test or caller agent is injected."""

    def __init__(self, *, model: str, config: dict[str, Any], max_tokens: int = 8192) -> None:
        """Bind provider config for summary calls."""
        self.model = model
        self.config = config
        self.max_tokens = max(512, int(max_tokens))

    def __call__(self, messages: list[dict[str, Any]], previous_summary: str | None = None) -> str:
        """Ask the configured provider to rewrite compactable messages."""
        from ..llms.registry import bare_model, detect_provider, get_provider_config

        provider_name = detect_provider(self.model)
        provider_config = get_provider_config(provider_name)
        model_name = bare_model(self.model)
        prompt = self._build_prompt(messages, previous_summary)

        if provider_config.type == "anthropic":
            return self._call_anthropic(provider_name, model_name, prompt)
        return self._call_openai_compatible(provider_name, model_name, prompt)

    @staticmethod
    def _build_prompt(messages: list[dict[str, Any]], previous_summary: str | None) -> str:
        """Build the compaction prompt sent to the provider."""
        prior = f"Existing summary to refresh:\n{previous_summary.strip()}\n\n" if previous_summary else ""
        return (
            "Rewrite the following conversation history into a compact, durable memory for the next model turn.\n"
            "Preserve concrete facts, user instructions, decisions, file paths, tool results, errors, fixes, "
            "open questions, and current task state. Drop chatter and duplicate text. Output only the summary.\n\n"
            f"{prior}"
            f"Conversation history to compact:\n{render_messages_for_summary(messages)}"
        )

    def _request_timeout(self) -> float:
        """Return provider request timeout for summary calls."""
        raw = self.config.get("llm_timeout") or self.config.get("request_timeout")
        try:
            return max(1.0, float(raw)) if raw is not None else 60.0
        except (TypeError, ValueError):
            return 60.0

    def _call_openai_compatible(self, provider_name: str, model_name: str, prompt: str) -> str:
        """Run a summary call against an OpenAI-compatible provider."""
        from openai import OpenAI

        from ..llms.registry import PROVIDERS, get_api_key, provider_default_headers

        provider = PROVIDERS.get(provider_name, PROVIDERS.get("openai"))
        api_key = self.config.get("api_key") or get_api_key(provider_name, self.config) or "dummy"
        base_url = (
            self.config.get("base_url")
            or self.config.get("custom_base_url")
            or (provider.base_url if provider else None)
            or "https://api.openai.com/v1"
        )
        client = OpenAI(
            api_key=api_key,
            base_url=base_url,
            default_headers=provider_default_headers(provider_name) or None,
            timeout=self._request_timeout(),
        )
        response = client.chat.completions.create(
            model=model_name,
            messages=[
                {
                    "role": "system",
                    "content": "You are a context compaction agent. Output only the rewritten summary.",
                },
                {"role": "user", "content": prompt},
            ],
            max_tokens=self.max_tokens,
            temperature=0.2,
        )
        return response.choices[0].message.content or ""

    def _call_anthropic(self, provider_name: str, model_name: str, prompt: str) -> str:
        """Run a summary call against Anthropic Messages."""
        import anthropic

        from ..llms.registry import get_api_key

        client = anthropic.Anthropic(
            api_key=self.config.get("api_key") or get_api_key(provider_name, self.config),
            timeout=self._request_timeout(),
        )
        response = client.messages.create(
            model=model_name,
            system="You are a context compaction agent. Output only the rewritten summary.",
            messages=[{"role": "user", "content": prompt}],
            max_tokens=self.max_tokens,
            temperature=0.2,
        )
        parts = []
        for block in response.content:
            text = getattr(block, "text", None)
            if text:
                parts.append(text)
        return "\n".join(parts)


def compaction_summary_agent_from_config(
    model: str,
    config: dict[str, Any],
) -> CompactionSummaryAgent | None:
    """Return an injected or provider-backed compaction summary agent."""
    injected = config.get("compaction_summary_agent")
    if callable(injected):
        return injected
    if injected is False:
        return None
    if not model:
        return None

    from ..llms.registry import PROVIDERS, detect_provider, get_api_key

    provider_name = detect_provider(model)
    provider = PROVIDERS.get(provider_name)
    explicit_base_url = bool(config.get("base_url") or config.get("custom_base_url"))
    api_key = config.get("api_key") or get_api_key(provider_name, config)

    if not api_key and provider_name not in _LOCAL_PROVIDER_NAMES and not explicit_base_url:
        return None
    if provider and provider.type == "anthropic" and not api_key:
        return None

    return ProviderCompactionAgent(
        model=model,
        config=config,
        max_tokens=int(config.get("compaction_summary_max_tokens") or 8192),
    )


class CompactionProvisioner:
    """Plan and execute agent-backed context compaction."""

    def __init__(
        self,
        *,
        model: str,
        max_context_tokens: int,
        summary_agent: CompactionSummaryAgent | None,
        threshold_tokens: int | None = None,
        target_tokens: int | None = None,
        threshold_ratio: float = _DEFAULT_THRESHOLD_RATIO,
        target_ratio: float = _DEFAULT_TARGET_RATIO,
        tail_ratio: float = _DEFAULT_TAIL_RATIO,
    ) -> None:
        """Configure token budgets and the summary agent."""
        self.model = model
        self.max_context_tokens = max(1, int(max_context_tokens or 1))
        self.threshold_tokens = threshold_tokens or int(self.max_context_tokens * threshold_ratio)
        self.target_tokens = target_tokens or int(self.max_context_tokens * target_ratio)
        self.threshold_tokens = max(1, self.threshold_tokens)
        self.target_tokens = max(1, min(self.target_tokens, self.threshold_tokens))
        self.tail_ratio = max(0.05, min(0.9, tail_ratio))
        self.summary_agent = summary_agent
        self.token_counter = SmartTokenCounter(model=model)

    def count_tokens(self, messages: list[dict[str, Any]]) -> int:
        """Return estimated tokens for ``messages``."""
        return self.token_counter.count_tokens(messages)

    def should_compact(
        self,
        messages: list[dict[str, Any]],
        *,
        appended_messages: list[dict[str, Any]] | None = None,
        force: bool = False,
    ) -> bool:
        """Return whether a compaction pass should run now."""
        if force:
            return True
        candidate = [*messages, *(appended_messages or [])]
        return self.count_tokens(candidate) >= self.threshold_tokens

    def compact_before_append(
        self,
        messages: list[dict[str, Any]],
        appended_messages: list[dict[str, Any]],
    ) -> CompactionProvision:
        """Compact existing context when appending new messages would overflow."""
        candidate = [*messages, *appended_messages]
        tokens_before = self.count_tokens(candidate)
        if tokens_before < self.threshold_tokens:
            return CompactionProvision(False, messages, tokens_before, tokens_before, reason="below_threshold")

        provision = self.compact(messages, force=True)
        if not provision.compacted:
            return CompactionProvision(
                False,
                messages,
                tokens_before,
                tokens_before,
                reason=provision.reason,
                error=provision.error,
            )
        return CompactionProvision(
            True,
            provision.messages,
            tokens_before,
            self.count_tokens([*provision.messages, *appended_messages]),
            summarized_count=provision.summarized_count,
            kept_count=provision.kept_count,
            reason=provision.reason,
        )

    def compact(
        self,
        messages: list[dict[str, Any]],
        *,
        force: bool = False,
        previous_summary: str | None = None,
    ) -> CompactionProvision:
        """Replace compactable history with an agent-written summary."""
        tokens_before = self.count_tokens(messages)
        if not force and tokens_before < self.threshold_tokens:
            return CompactionProvision(False, messages, tokens_before, tokens_before, reason="below_threshold")

        if self.summary_agent is None:
            return CompactionProvision(False, messages, tokens_before, tokens_before, reason="no_summary_agent")

        partition = self._partition(messages)
        if partition is None:
            return CompactionProvision(False, messages, tokens_before, tokens_before, reason="nothing_to_compact")

        system_messages, compactable_messages, live_messages = partition
        try:
            summary = self.summary_agent(compactable_messages, previous_summary).strip()
        except Exception as exc:
            logger.warning("Context compaction agent failed", exc_info=True)
            return CompactionProvision(
                False,
                messages,
                tokens_before,
                tokens_before,
                reason="summary_agent_failed",
                error=str(exc),
            )

        if not summary:
            return CompactionProvision(False, messages, tokens_before, tokens_before, reason="empty_summary")

        summary_message = {
            "role": "user",
            "content": f"{COMPACTION_SUMMARY_PREFIX} - {len(compactable_messages)} messages compacted]\n\n{summary}",
        }
        compacted = repair_tool_message_sequence([*system_messages, summary_message, *live_messages])
        tokens_after = self.count_tokens(compacted)
        if tokens_after >= tokens_before:
            return CompactionProvision(False, messages, tokens_before, tokens_after, reason="summary_did_not_shrink")

        return CompactionProvision(
            True,
            compacted,
            tokens_before,
            tokens_after,
            summarized_count=len(compactable_messages),
            kept_count=len(live_messages),
            reason="compacted",
        )

    def _partition(
        self,
        messages: list[dict[str, Any]],
    ) -> tuple[list[dict[str, Any]], list[dict[str, Any]], list[dict[str, Any]]] | None:
        """Split messages into system, compactable, and live suffix by token budget."""
        system_messages = [message for message in messages if message.get("role") == "system"]
        conversation_messages = [message for message in messages if message.get("role") != "system"]
        if len(conversation_messages) < 2:
            return None

        tail_budget = max(1, int(self.target_tokens * self.tail_ratio))
        live_messages: list[dict[str, Any]] = []
        live_tokens = 0
        for message in reversed(conversation_messages):
            message_tokens = max(1, self.count_tokens([message]))
            if live_messages and live_tokens + message_tokens > tail_budget:
                break
            live_messages.insert(0, message)
            live_tokens += message_tokens

        if not live_messages:
            live_messages = [conversation_messages[-1]]

        start = len(conversation_messages) - len(live_messages)
        while start > 0 and conversation_messages[start].get("role") == "tool":
            start -= 1
            if conversation_messages[start].get("role") == "assistant":
                break

        compactable_messages = conversation_messages[:start]
        live_messages = conversation_messages[start:]
        if not compactable_messages or not live_messages:
            return None
        return system_messages, compactable_messages, live_messages

File: src/typescript/src/context/compactionProvisioner.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import {
  COMPACTION_REFERENCE_PREFIX,
  ContextCompressor,
  type ContextMessage,
  type Summarizer,
} from './compressor.js'
import { SmartTokenCounter } from './tokenCounter.js'
import { repairToolMessageSequence } from './toolPairRepair.js'

export { repairToolMessageSequence } from './toolPairRepair.js'

/** Prefix emitted by the shared compressor for provisioned summary messages. */
export const COMPACTION_SUMMARY_PREFIX = COMPACTION_REFERENCE_PREFIX
export const DEFAULT_COMPACTION_TAIL_RATIO = 0.35
export const DEFAULT_COMPACTION_TARGET_RATIO = 0.5
export const DEFAULT_COMPACTION_THRESHOLD_RATIO = 0.75
export const DEFAULT_COMPACTION_SUMMARY_MAX_TOKENS = 8_192
export const DEFAULT_COMPACTION_SUMMARY_TEMPERATURE = 0.2

/** Caller-supplied summary function; no provider client is constructed by context code. */
export type CompactionSummaryAgent = (
  messages: readonly ContextMessage[],
  previousSummary?: string,
) => string

/** Fully resolved request delivered to an injected model port. */
export interface CompactionModelRequest {
  readonly maxTokens: number
  readonly messages: readonly ContextMessage[]
  readonly model: string
  readonly previousSummary?: string
  readonly prompt: string
  readonly temperature: number
}

/** Model boundary for summary generation. It is intentionally synchronous to match ContextCompressor. */
export type CompactionModelPort = (request: CompactionModelRequest) => string

export interface ProviderCompactionAgentOptions {
  readonly maxTokens?: number
  readonly model: string
  readonly modelPort: CompactionModelPort
  readonly temperature?: number
}

/**
 * Adapter for a caller-owned model port.
 *
 * Despite the historical Python name, this class does not discover providers,
 * read credentials, or create network clients. The host owns those concerns.
 */
export class ProviderCompactionAgent {
  readonly maxTokens: number
  readonly model: string
  readonly temperature: number
  private readonly modelPort: CompactionModelPort

  constructor(options: ProviderCompactionAgentOptions) {
    this.model = requireText(options.model, 'model')
    if (typeof options.modelPort !== 'function') {
      throw new TypeError('modelPort must be a function')
    }
    this.modelPort = options.modelPort
    const requestedMaxTokens = positiveInteger(
      options.maxTokens ?? DEFAULT_COMPACTION_SUMMARY_MAX_TOKENS,
      'maxTokens',
    )
    this.maxTokens = Math.max(512, requestedMaxTokens)
    this.temperature = nonNegativeFiniteNumber(
      options.temperature ?? DEFAULT_COMPACTION_SUMMARY_TEMPERATURE,
      'temperature',
    )
  }

  summarize(messages: readonly ContextMessage[], previousSummary?: string): string {
    const request: CompactionModelRequest = {
      model: this.model,
      messages,
      prompt: buildCompactionPrompt(messages, previousSummary),
      maxTokens: this.maxTokens,
      temperature: this.temperature,
      ...(previousSummary === undefined ? {} : { previousSummary }),
    }
    const summary = this.modelPort(request)
    if (typeof summary !== 'string') {
      throw new TypeError('compaction model port must return a string')
    }
    return summary
  }

  toSummaryAgent(): CompactionSummaryAgent {
    return (messages, previousSummary) => this.summarize(messages, previousSummary)
  }
}

/** Create a summary agent from an explicit host-owned model invocation port. */
export function compactionSummaryAgentFromModelPort(
  model: string,
  modelPort: CompactionModelPort,
  options: Omit<ProviderCompactionAgentOptions, 'model' | 'modelPort'> = {},
): CompactionSummaryAgent {
  return new ProviderCompactionAgent({ model, modelPort, ...options }).toSummaryAgent()
}

export interface CompactionProvision {
  readonly compacted: boolean
  readonly error: string
  readonly keptCount: number
  readonly messages: ContextMessage[]
  readonly reason: string
  readonly summarizedCount: number
  readonly tokensAfter: number
  readonly tokensBefore: number
}

export interface CompactionProvisionerOptions {
  readonly maxContextTokens: number
  readonly model: string
  readonly modelPort?: CompactionModelPort
  readonly summaryAgent?: CompactionSummaryAgent
  readonly summaryMaxTokens?: number
  readonly summaryTemperature?: number
  readonly tailRatio?: number
  readonly targetRatio?: number
  readonly targetTokens?: number
  readonly thresholdRatio?: number
  readonly thresholdTokens?: number
  readonly tokenCounter?: SmartTokenCounter
}

export interface CompactionDecisionOptions {
  readonly appendedMessages?: readonly ContextMessage[]
  readonly force?: boolean
}

export interface CompactOptions {
  readonly force?: boolean
  readonly previousSummary?: string
}

/**
 * Plan and execute a model-backed compaction pass around the shared compressor.
 *
 * Thresholding and the protected live tail follow the Python provisioner. The
 * actual pruning, summary placement, iterative-summary handling, and token
 * accounting stay in ContextCompressor so the runtime has one compressor.
 */
export class CompactionProvisioner {
  readonly maxContextTokens: number
  readonly model: string
  readonly tailRatio: number
  readonly targetTokens: number
  readonly thresholdTokens: number
  private readonly summaryAgent: CompactionSummaryAgent | undefined
  private readonly tokenCounter: SmartTokenCounter

  constructor(options: CompactionProvisionerOptions) {
    this.model = requireText(options.model, 'model')
    this.maxContextTokens = positiveInteger(options.maxContextTokens, 'maxContextTokens')
    const thresholdRatio = nonNegativeFiniteNumber(
      options.thresholdRatio ?? DEFAULT_COMPACTION_THRESHOLD_RATIO,
      'thresholdRatio',
    )
    const targetRatio = nonNegativeFiniteNumber(
      options.targetRatio ?? DEFAULT_COMPACTION_TARGET_RATIO,
      'targetRatio',
    )
    const configuredThreshold = options.thresholdTokens === undefined
      ? Math.floor(this.maxContextTokens * thresholdRatio)
      : positiveInteger(options.thresholdTokens, 'thresholdTokens')
    const configuredTarget = options.targetTokens === undefined
      ? Math.floor(this.maxContextTokens * targetRatio)
      : positiveInteger(options.targetTokens, 'targetTokens')
    this.thresholdTokens = Math.max(1, configuredThreshold)
    this.targetTokens = Math.max(1, Math.min(configuredTarget, this.thresholdTokens))
    this.tailRatio = clamp(
      nonNegativeFiniteNumber(options.tailRatio ?? DEFAULT_COMPACTION_TAIL_RATIO, 'tailRatio'),
      0.05,
      0.9,
    )
    if (options.summaryAgent !== undefined && typeof options.summaryAgent !== 'function') {
      throw new TypeError('summaryAgent must be a function')
    }
    if (options.modelPort !== undefined && typeof options.modelPort !== 'function') {
      throw new TypeError('modelPort must be a function')
    }
    this.summaryAgent = options.summaryAgent ?? (options.modelPort === undefined
      ? undefined
      : compactionSummaryAgentFromModelPort(this.model, options.modelPort, {
        ...(options.summaryMaxTokens === undefined ? {} : { maxTokens: options.summaryMaxTokens }),
        ...(options.summaryTemperature === undefined ? {} : { temperature: options.summaryTemperature }),
      }))
    this.tokenCounter = options.tokenCounter ?? new SmartTokenCounter({ model: this.model })
  }

  /** Return the shared model-aware estimate for a message sequence. */
  countTokens(messages: readonly ContextMessage[]): number {
    return this.tokenCounter.countTokens(messages)
  }

  /** Decide against the full post-append context, without mutating either input. */
  shouldCompact(messages: readonly ContextMessage[], options: CompactionDecisionOptions = {}): boolean {
    if (options.force) return true
    const candidate = [...messages, ...(options.appendedMessages ?? [])]
    return this.countTokens(candidate) >= this.thresholdTokens
  }

  /** Compact only existing history before an incoming turn is appended. */
  compactBeforeAppend(
    messages: readonly ContextMessage[],
    appendedMessages: readonly ContextMessage[],
  ): CompactionProvision {
    const candidate = [...messages, ...appendedMessages]
    const tokensBefore = this.countTokens(candidate)
    if (tokensBefore < this.thresholdTokens) {
      return unchanged(messages, tokensBefore, 'below_threshold')
    }

    const provision = this.compact(messages, { force: true })
    if (!provision.compacted) {
      return {
        ...unchanged(messages, tokensBefore, provision.reason),
        error: provision.error,
      }
    }
    return {
      ...provision,
      tokensBefore,
      tokensAfter: this.countTokens([...provision.messages, ...appendedMessages]),
    }
  }

  /** Replace compactable history with a model-written reference summary. */
  compact(messages: readonly ContextMessage[], options: CompactOptions = {}): CompactionProvision {
    const original = [...messages]
    const tokensBefore = this.countTokens(original)
    if (!options.force && tokensBefore < this.thresholdTokens) {
      return unchanged(original, tokensBefore, 'below_threshold')
    }
    if (this.summaryAgent === undefined) {
      return unchanged(original, tokensBefore, 'no_summary_agent')
    }

    const systemMessages = original.filter(message => message.role === 'system')
    const conversationMessages = original.filter(message => message.role !== 'system')
    const protectedTail = this.protectedTailCount(conversationMessages)
    if (conversationMessages.length < 2 || protectedTail >= conversationMessages.length) {
      return unchanged(original, tokensBefore, 'nothing_to_compact')
    }

    let summaryCalled = false
    let summaryFailure: unknown
    let summaryWasEmpty = false
    const compressor = this.compressor(protectedTail, (compactableMessages, _budgetTokens) => {
      summaryCalled = true
      try {
        const summary = options.previousSummary === undefined
          ? this.summaryAgent!(compactableMessages)
          : this.summaryAgent!(compactableMessages, options.previousSummary)
        if (typeof summary !== 'string') {
          throw new TypeError('summaryAgent must return a string')
        }
        const trimmed = summary.trim()
        if (!trimmed) {
          summaryWasEmpty = true
          return ''
        }
        return trimmed
      } catch (error) {
        summaryFailure = error
        return ''
      }
    })
    const compressed = compressor.compress(conversationMessages)
    if (summaryFailure !== undefined) {
      return failure(original, tokensBefore, 'summary_agent_failed', summaryFailure)
    }
    if (summaryWasEmpty) {
      return unchanged(original, tokensBefore, 'empty_summary')
    }
    if (!compressed.compressed) {
      return unchanged(original, tokensBefore, 'nothing_to_compact')
    }

    const output = repairToolMessageSequence([...systemMessages, ...compressed.messages])
    const tokensAfter = this.countTokens(output)
    if (tokensAfter >= tokensBefore) {
      return {
        ...unchanged(original, tokensBefore, 'summary_did_not_shrink'),
        tokensAfter,
      }
    }
    return {
      compacted: true,
      messages: output,
      tokensBefore,
      tokensAfter,
      summarizedCount: compressed.compressedCount,
      keptCount: protectedTail,
      reason: summaryCalled ? 'compacted' : 'pruned',
      error: '',
    }
  }

  private compressor(protectLast: number, summarizer: Summarizer): ContextCompressor {
    return new ContextCompressor({
      contextWindow: this.maxContextTokens,
      model: this.model,
      protectFirst: 0,
      protectLast,
      summarizer,
      summaryMaxTokens: this.targetTokens,
      summaryMinTokens: 1,
      threshold: clamp(this.thresholdTokens / this.maxContextTokens, Number.EPSILON, 1),
      tokenCounter: this.tokenCounter,
    })
  }

  private protectedTailCount(messages: readonly ContextMessage[]): number {
    if (!messages.length) return 0
    const tailBudget = Math.max(1, Math.floor(this.targetTokens * this.tailRatio))
    let start = messages.length
    let tailTokens = 0
    for (let index = messages.length - 1; index >= 0; index -= 1) {
      const message = messages[index]
      if (message === undefined) continue
      const messageTokens = Math.max(1, this.countTokens([message]))
      if (start !== messages.length && tailTokens + messageTokens > tailBudget) break
      start = index
      tailTokens += messageTokens
    }
    if (start === messages.length) start = messages.length - 1
    while (start > 0 && messages[start]?.role === 'tool') {
      start -= 1
      if (messages[start]?.role === 'assistant') break
    }
    return messages.length - start
  }
}

/** Render provider-shaped message content as readable text for a summary prompt. */
export function messageContentToText(content: unknown): string {
  if (typeof content === 'string') return content
  if (Array.isArray(content)) {
    return content.map(item => isRecord(item)
      ? typeof item.text === 'string' ? item.text : stableJson(item)
      : String(item)).join('\n')
  }
  if (isRecord(content)) return stableJson(content)
  return content === undefined || content === null ? '' : String(content)
}

/** Render full messages into the explicit, deterministic prompt input supplied to a model port. */
export function renderMessagesForSummary(messages: readonly ContextMessage[]): string {
  return messages.map((message, index) => {
    const role = typeof message.role === 'string' ? message.role.toUpperCase() : 'UNKNOWN'
    const lines = [`Message ${index + 1} [${role}]`]
    const content = messageContentToText(message.content)
    if (content) lines.push(content)
    if (message.tool_calls) lines.push(`tool_calls=${stableJson(message.tool_calls)}`)
    if (message.tool_call_id) lines.push(`tool_call_id=${String(message.tool_call_id)}`)
    return lines.join('\n')
  }).join('\n\n')
}

/** Build the durable-memory instruction used by ProviderCompactionAgent's injected model port. */
export function buildCompactionPrompt(messages: readonly ContextMessage[], previousSummary?: string): string {
  const prior = previousSummary?.trim()
  return [
    'Rewrite the following conversation history into a compact, durable memory for the next model turn.',
    'Preserve concrete facts, user instructions, decisions, file paths, tool results, errors, fixes,',
    'open questions, and current task state. Drop chatter and duplicate text. Output only the summary.',
    '',
    ...(prior ? ['Existing summary to refresh:', prior, ''] : []),
    'Conversation history to compact:',
    renderMessagesForSummary(messages),
  ].join('\n')
}

function unchanged(messages: readonly ContextMessage[], tokens: number, reason: string): CompactionProvision {
  return {
    compacted: false,
    messages: [...messages],
    tokensBefore: tokens,
    tokensAfter: tokens,
    summarizedCount: 0,
    keptCount: 0,
    reason,
    error: '',
  }
}

function failure(
  messages: readonly ContextMessage[],
  tokens: number,
  reason: string,
  error: unknown,
): CompactionProvision {
  return { ...unchanged(messages, tokens, reason), error: errorMessage(error) }
}

function stableJson(value: unknown): string {
  try {
    return JSON.stringify(value, (_key, item: unknown) => {
      if (typeof item === 'bigint') return item.toString()
      if (!isRecord(item)) return item
      return Object.fromEntries(Object.keys(item).sort().map(key => [key, item[key]]))
    }) ?? String(value)
  } catch {
    return String(value)
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function requireText(value: string, label: string): string {
  if (typeof value !== 'string' || !value.trim()) throw new TypeError(`${label} must be a non-empty string`)
  return value
}

function positiveInteger(value: number, label: string): number {
  if (!Number.isSafeInteger(value) || value < 1) throw new RangeError(`${label} must be a positive integer`)
  return value
}

function nonNegativeFiniteNumber(value: number, label: string): number {
  if (!Number.isFinite(value) || value < 0) throw new RangeError(`${label} must be a non-negative finite number`)
  return value
}

function clamp(value: number, lower: number, upper: number): number {
  return Math.max(lower, Math.min(upper, value))
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error)
}
